### PR TITLE
Backend overhaul

### PR DESCRIPTION
**Pull request by Sefiraat**
This is to fix my bad code as much as possible without changing anything breaking.

- [x] Remove Lombok
- [x] Tidy up some general crap
- [x] Reduce code smells where possible
- [x] Convert CM classes to builders where appropriate
- [x] Move events inside traits directly

and anything else to un-dumb.

This will be followed by a second pull to convert all tools to proper PDC Types which will need a much longer lead in time.

### CODE REVIEW

**Comment by ybw0014** [comment]
Wow, BIG change

File: src/main/java/io/github/sefiraat/slimetinker/items/componentmaterials/CMManager.java [hunk 1]
```diff
@@ -142,22 +129,22 @@ public CMManager() {
             String id = entry.getKey();
 
             // Tools, armour and kits (referenced through dummy)
-            if (cm.isValidToolRod() & traitManager.isEnabled(id, IDStrings.ROD)) {
+            if (cm.isValidToolRod() && traitManager.isEnabled(id, IDStrings.ROD)) {
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/componentmaterials/CMManager.java:132]: how did I...

Comment (Sefiraat) [reply src/main/java/io/github/sefiraat/slimetinker/items/componentmaterials/CMManager.java:132]: I've no idea - luckily it just means a bit of extra overhead - and it meant I learned how to bypass short-circuiting :D!

Comment (Sefiraat) [reply src/main/java/io/github/sefiraat/slimetinker/items/componentmaterials/CMManager.java:132]: Also, I didn't see them during the review soooooooooo 🥇

Comment (ybw0014) [reply src/main/java/io/github/sefiraat/slimetinker/items/componentmaterials/CMManager.java:132]: I made this change, but everything except bulky works fine.

**Comment by Sefiraat** [comment]
This is a bit of a massive PR, but I would love you to take a look over it if you have time. If you don't - please do not worry and, if you do, please do take your time, there is no rush :)

File: src/main/java/io/github/sefiraat/slimetinker/items/templates/ArmourDefinition.java [hunk 2]
```diff
@@ -1,16 +1,39 @@
...
 
+    public ArmourDefinition(String classType, String partType, String plateMaterial, String gambesonMaterial, String linksMaterial) {
+        this.classType = classType;
+        this.partType = partType;
+        this.plateMaterial = plateMaterial;
+        this.gambesonMaterial = gambesonMaterial;
+        this.linksMaterial = linksMaterial;
+    }
+
+    private final String classType;
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/templates/ArmourDefinition.java:18]: why it is below constructor?

File: src/main/java/io/github/sefiraat/slimetinker/items/templates/ArmourTemplate.java [hunk 3]
```diff
@@ -25,22 +25,22 @@ public ArmourTemplate(ItemGroup itemGroup, SlimefunItemStack item, RecipeType re
 
     public String getName(ArmourDefinition armourDefinition) {
         return
-            CMManager.getById(armourDefinition.getPlateMaterial()).getColor() + ThemeUtils.toTitleCase(armourDefinition.getPlateMaterial()) + "-" +
-                CMManager.getById(armourDefinition.getGambesonMaterial()).getColor() + ThemeUtils.toTitleCase(armourDefinition.getGambesonMaterial()) + "-" +
-                CMManager.getById(armourDefinition.getLinksMaterial()).getColor() + ThemeUtils.toTitleCase(armourDefinition.getLinksMaterial()) + " " +
+            TinkerMaterialManager.getById(armourDefinition.getPlateMaterial()).getColor() + ThemeUtils.toTitleCase(armourDefinition.getPlateMaterial()) + "-" +
+                TinkerMaterialManager.getById(armourDefinition.getGambesonMaterial()).getColor() + ThemeUtils.toTitleCase(armourDefinition.getGambesonMaterial()) + "-" +
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/templates/ArmourTemplate.java:29]: Should the indent here same as above?

File: src/main/java/io/github/sefiraat/slimetinker/items/templates/ToolTemplate.java [hunk 4]
```diff
@@ -26,24 +26,24 @@ public ToolTemplate(ItemGroup itemGroup, SlimefunItemStack item, RecipeType reci
 
     public String getName(ToolDefinition toolDefinition) {
         return
-            CMManager.getById(toolDefinition.getHeadMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getHeadMaterial()) + "-" +
-                CMManager.getById(toolDefinition.getBinderMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getBinderMaterial()) + "-" +
-                CMManager.getById(toolDefinition.getRodMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getRodMaterial()) + " " +
+            TinkerMaterialManager.getById(toolDefinition.getHeadMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getHeadMaterial()) + "-" +
+                TinkerMaterialManager.getById(toolDefinition.getBinderMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getBinderMaterial()) + "-" +
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/templates/ToolTemplate.java:30]: indent

File: src/main/java/io/github/sefiraat/slimetinker/items/templates/ToolTemplateExplosive.java [hunk 5]
```diff
@@ -27,31 +27,31 @@ public ToolTemplateExplosive(ItemGroup itemGroup, SlimefunItemStack item, Recipe
...
             );
     }
 
     public String getName(ToolDefinition toolDefinition) {
         return
-            CMManager.getById(toolDefinition.getHeadMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getHeadMaterial()) + "-" +
-                CMManager.getById(toolDefinition.getBinderMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getBinderMaterial()) + "-" +
-                CMManager.getById(toolDefinition.getRodMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getRodMaterial()) + " " +
+            TinkerMaterialManager.getById(toolDefinition.getHeadMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getHeadMaterial()) + "-" +
+                TinkerMaterialManager.getById(toolDefinition.getBinderMaterial()).getColor() + ThemeUtils.toTitleCase(toolDefinition.getBinderMaterial()) + "-" +
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/templates/ToolTemplateExplosive.java:38]: indent

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/TinkerMaterial.java [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+            this.alloy.setupAlloy(this);
+        }
+
+        setupTraits();
+        registerParts();
+        return this;
+    }
+
+    public void setupTraits() {
+        final TraitManager manager = SlimeTinker.inst().getTraitManager();
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/TinkerMaterial.java:112]:
```suggestion
        final TraitManager manager = SlimeTinker.getInstance().getTraitManager();
```

File: src/main/java/io/github/sefiraat/slimetinker/SlimeTinker.java [hunk 7]
```diff
@@ -85,4 +73,36 @@ protected void disable() {
         instance = null;
     }
 
+    public static SlimeTinker getInstance() {
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/SlimeTinker.java:76]: there is another instance getter `inst()`, either remove this or that one

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/TinkerMaterial.java [hunk 8]
```diff
@@ -0,0 +1,655 @@
...
+            this.traitArmorGambeson.setupTrait(this);
+        }
+        if (this.traitArmorLinks != null && manager.getEnabled(this.id, Ids.LINKS)) {
+            this.traitArmorLinks.setupTrait(this);
+        }
+        manager.save();
+    }
+
+    public void registerParts() {
+        TraitManager traitManager = SlimeTinker.inst().getTraitManager();
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/TinkerMaterial.java:137]:
```suggestion
        TraitManager traitManager = SlimeTinker.getInstance().getTraitManager();
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/TinkerMaterial.java [hunk 9]
```diff
@@ -0,0 +1,655 @@
...
+
+        // Heads (and repair kits)
+        if (this.traitToolHead != null && traitManager.isEnabled(this.id, Ids.HEAD)) {
+            new PartTemplate(
+                ItemGroups.PART_DICT,
+                headStack(this.id, "SWORD", SkullTextures.PART_SWORD_BLADE),
+                DummySmeltery.TYPE,
+                basicRecipe(Casts.CAST_SWORDBLADE, getLiquidItemStack(2)),
+                this.id
+            ).register(SlimeTinker.inst());
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/TinkerMaterial.java:147]: these inst()

**Review by ybw0014 — CHANGES_REQUESTED** [review]
haven't reviewed all

**Comment by ybw0014** [comment]
reviewed 51 files, will review remaining files later

File: src/main/java/io/github/sefiraat/slimetinker/utils/Keys.java [hunk 10]
```diff
@@ -1,82 +1,83 @@
...
-    private final NamespacedKey itemGroupAlloys = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-alloys");
-    private final NamespacedKey itemGroupProperties = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-properties");
-    private final NamespacedKey itemGroupCasts = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-casts");
-    private final NamespacedKey itemGroupParts = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-parts");
-    private final NamespacedKey itemGroupPartDict = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-part-dict");
-    private final NamespacedKey itemGroupTools = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-tools");
-    private final NamespacedKey itemGroupArmour = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-armour");
-    private final NamespacedKey itemGroupModifications = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-modifications");
-    private final NamespacedKey itemGroupDummy = new NamespacedKey(SlimeTinker.inst(), "slime-tinker-dummy");
+    public static final NamespacedKey ITEM_GROUP_MAIN = new NamespacedKey(SlimeTinker.inst(), "slime-tinker");
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/utils/Keys.java]: `inst()` thing

File: src/main/java/io/github/sefiraat/slimetinker/utils/ItemUtils.java [hunk 11]
```diff
@@ -284,30 +279,29 @@ private static void rebuildArmourName(ItemStack itemStack) {
...
+    private static void setName(@Nonnull ItemStack itemStack, String first, String second, String third, String type) {
         ItemMeta im = itemStack.getItemMeta();
         assert im != null;
 
         String name =
-            CMManager.getById(first).getColor() + ThemeUtils.toTitleCase(first) + "-" +
-                CMManager.getById(second).getColor() + ThemeUtils.toTitleCase(second) + "-" +
-                CMManager.getById(third).getColor() + ThemeUtils.toTitleCase(third) + " " +
+            TinkerMaterialManager.getById(first).getColor() + ThemeUtils.toTitleCase(first) + "-" +
+                TinkerMaterialManager.getById(second).getColor() + ThemeUtils.toTitleCase(second) + "-" +
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/utils/ItemUtils.java:290]: indent

**Review by ybw0014 — COMMENTED** [review]
more reviews

File: src/main/java/io/github/sefiraat/slimetinker/items/templates/RepairkitTemplate.java [hunk 12]
```diff
@@ -31,11 +29,10 @@ public RepairkitTemplate(ItemGroup itemGroup, SlimefunItemStack item, RecipeType
     }
 
     public static boolean isRepairKit(ItemStack itemStack) {
-        NamespacedKey key = SlimeTinker.inst().getKeys().getPartClass();
+        NamespacedKey key = Keys.PART_CLASS;
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/templates/RepairkitTemplate.java:32]: maybe put this in `PersistentDataAPI.getString()` directly

**Review by ybw0014 — COMMENTED** [review]
will review traits when I get up

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 13]
```diff
@@ -0,0 +1,2620 @@
...
+            "25% chance to afflict poison on hit."
+        )
+        .addConsumer(TraitEventType.ENTITY_DAMAGED, EntityDamageEvents::headLead);
+
+    public static final MaterialTrait CORE_LEAD_ROD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_ROD)
+        .setAddedBy(SupportedPluginsManager.CORE_NOTE)
+        .setTraitName("Leech")
+        .setLore(
+            "Bonus health) but drains your energy."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Bonus health, but drains your energy."
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 14]
```diff
@@ -0,0 +1,2620 @@
...
+        )
+        .addConsumer(TraitEventType.TICK, TickEvents::plateFerrosilicon);
+
+    public static final MaterialTrait CORE_FERROSILICON_LINKS = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_LINKS)
+        .setAddedBy(SupportedPluginsManager.CORE_NOTE)
+        .setTraitName("Brightburn")
+        .setLore(
+            "Gives on stack of Burning Bright! With 4 ",
+            "stacks) hostile mobs are repelled (not bosses)."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "stacks, hostile mobs are repelled (not bosses)."
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 15]
```diff
@@ -0,0 +1,2620 @@
...
+            "The void is no longer your enemy"
+        )
+        .addConsumer(TraitEventType.PLAYER_DAMAGED, PlayerDamagedEvents::plateVoid);
+
+    public static final MaterialTrait INFINITY_VOID_LINKS = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_LINKS)
+        .setAddedBy(SupportedPluginsManager.INFINITY_EXPANSION_NOTE)
+        .setTraitName("The End")
+        .setLore(
+            "Reduce damage from the Dragon) Endermen and",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Reduce damage from the Dragon, Endermen and",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 16]
```diff
@@ -0,0 +1,2620 @@
...
+            "Chance to restore durability on hit/block break"
+        )
+        .addConsumer(TraitEventType.DURABILITY, DurabilityEvents::headSingAluminum);
+
+    public static final MaterialTrait INFINITY_ALUMINUM_SINGULARITY_ROD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_ROD)
+        .setAddedBy(SupportedPluginsManager.INFINITY_EXPANSION_NOTE)
+        .setTraitName("Soft II")
+        .setLore(
+            "Durability loss +100%) tool EXP +100%."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Durability loss +100%, tool EXP +100%."
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 17]
```diff
@@ -0,0 +1,2620 @@
...
+            "within a 10 block range"
+        )
+        .addConsumer(TraitEventType.TICK, TickEvents::rodSingMagnesium);
+
+    public static final MaterialTrait INFINITY_MAGNESIUM_SINGULARITY_PLATES = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_PLATES)
+        .setAddedBy(SupportedPluginsManager.INFINITY_EXPANSION_NOTE)
+        .setTraitName("Light II")
+        .setLore(
+            "Damage Dealt -50%. Speed + 2"
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Damage dealt -50%. Speed + 2"
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 18]
```diff
@@ -0,0 +1,2620 @@
...
+            "All damage is halved (rounded up)."
+        )
+        .addConsumer(TraitEventType.PLAYER_DAMAGED, PlayerDamagedEvents::rodInfinity);
+
+    public static final MaterialTrait INFINITY_INFINITY_PLATES = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_PLATES)
+        .setAddedBy(SupportedPluginsManager.INFINITY_EXPANSION_NOTE)
+        .setTraitName("Infinite Capacity")
+        .setLore(
+            "Everytime you take damage) 10% is stored",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Every time you take damage, 10% is stored,",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 19]
```diff
@@ -0,0 +1,2620 @@
...
+            "Does nothing. Perhaps if you level it up?."
+        )
+        .addConsumer(TraitEventType.ENTITY_DAMAGED, EntityDamageEvents::headRefinedIron);
+
+    public static final MaterialTrait LITE_REFINED_IRON_ROD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_ROD)
+        .setAddedBy(SupportedPluginsManager.LITEXPANSION_NOTE)
+        .setTraitName("Manners Maketh Man")
+        .setLore(
+            "Speed 1) Damage + 50%. Your fights look",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Speed 1, Damage + 50%. Your fights look",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 20]
```diff
@@ -0,0 +1,2620 @@
...
+            "Cleave damage does not trigger Tinker's effects."
+        )
+        .addConsumer(TraitEventType.ENTITY_DAMAGED, EntityDamageEvents::headAdvancedAlloy);
+
+    public static final MaterialTrait LITE_ADVANCED_ALLOY_ROD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_ROD)
+        .setAddedBy(SupportedPluginsManager.LITEXPANSION_NOTE)
+        .setTraitName("Heavy Duty")
+        .setLore(
+            "When durability gets low) it eats 1 iron",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "When durability gets low, it eats 1 iron",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 21]
```diff
@@ -0,0 +1,2620 @@
...
+        )
+        .addConsumer(TraitEventType.PLAYER_DAMAGED, PlayerDamagedEvents::plateScrap)
+        .addConsumer(TraitEventType.ENTITY_DAMAGED, EntityDamageEvents::plateScrap);
+
+    public static final MaterialTrait LITE_SCRAP_LINKS = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_LINKS)
+        .setAddedBy(SupportedPluginsManager.LITEXPANSION_NOTE)
+        .setTraitName("Falling Apart")
+        .setLore(
+            "Exp + 200%. Chance for the amour to",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Exp + 200%. Chance for the armour to",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 22]
```diff
@@ -0,0 +1,2620 @@
...
+            "Magic Damage -50%"
+        )
+        .addConsumer(TraitEventType.PLAYER_DAMAGED, PlayerDamagedEvents::plateAluBrass);
+
+    public static final MaterialTrait CORE_ALUMINUM_BRASS_LINKS = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_LINKS)
+        .setAddedBy(SupportedPluginsManager.CORE_NOTE)
+        .setTraitName("Escape")
+        .setLore(
+            "Randomly teleport then hit (10s CD)"
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]: Consider having a consistent cd format. 
```suggestion
            "Randomly teleport then hit. (CD: 10s)"
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 23]
```diff
@@ -0,0 +1,2620 @@
...
+            "Smelts things when possible."
+        )
+        .addConsumer(TraitEventType.BLOCK_BREAK, BlockBreakEvents::headCorbronze);
+
+    public static final MaterialTrait INFINITY_MAGNONIUM_PLATES = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_PLATES)
+        .setAddedBy(SupportedPluginsManager.INFINITY_EXPANSION_NOTE)
+        .setTraitName("Escape")
+        .setLore(
+            "When hit, teleport away randomly (10s CD)."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "When hit, teleport away randomly. (CD: 10s)"
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 24]
```diff
@@ -0,0 +1,2620 @@
...
+            "Lava turns into magma blocks beneath your feet."
+        )
+        .addConsumer(TraitEventType.TICK, TickEvents::plateEarth);
+
+    public static final MaterialTrait INFINITY_EARTH_SINGULARITY_LINKS = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_LINKS)
+        .setAddedBy(SupportedPluginsManager.INFINITY_EXPANSION_NOTE)
+        .setTraitName("It's only natural")
+        .setLore(
+            "Feeds two nearby animals on right click. 2m CD",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Feeds two nearby animals on right click. (CD: 2m)",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 25]
```diff
@@ -0,0 +1,2620 @@
...
+        )
+        .addConsumer(TraitEventType.ENTITY_DAMAGED, EntityDamageEvents::linksScrap)
+        .addConsumer(TraitEventType.PLAYER_DAMAGED, PlayerDamagedEvents::linksScrap);
+
+    public static final MaterialTrait LITE_IRIDIUM_HEAD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_HEAD)
+        .setAddedBy(SupportedPluginsManager.LITEXPANSION_NOTE)
+        .setTraitName("Protective")
+        .setLore(
+            "Saves you from deaths. 20min CD."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Saves you from death. (CD: 20m)"
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 26]
```diff
@@ -0,0 +1,2620 @@
...
+        )
+        .addConsumer(TraitEventType.PLAYER_DAMAGED, PlayerDamagedEvents::headIridium);
+
+    public static final MaterialTrait LITE_IRIDIUM_ROD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_ROD)
+        .setAddedBy(SupportedPluginsManager.LITEXPANSION_NOTE)
+        .setTraitName("Warp")
+        .setLore(
+            "Turns your opponent around. 20s CD."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Turns your opponent around. (CD: 20s)"
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 27]
```diff
@@ -0,0 +1,2620 @@
...
+        )
+        .addConsumer(TraitEventType.ENTITY_DAMAGED, EntityDamageEvents::rodOsmium);
+
+    public static final MaterialTrait SFW_OSMIUM_PLATES = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_PLATES)
+        .setAddedBy(SupportedPluginsManager.SLIMEFUN_WARFARE_NOTE)
+        .setTraitName("Increased Mass")
+        .setLore(
+            "Attackers will be knocked back.",
+            "Gives you slow when procc'd"
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Gives you slow when procced"
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 28]
```diff
@@ -0,0 +1,2620 @@
...
+        .addConsumer(TraitEventType.PLAYER_DAMAGED, PlayerDamagedEvents::bindGhostly);
+
+    public static final MaterialTrait DYN_GHOSTLY_ESSENCE_ROD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_ROD)
+        .setAddedBy(SupportedPluginsManager.DYNATECH_NOTE)
+        .setTraitName("HyperCube [B]")
+        .setLore(
+            "Only works with both [A] and [B]. Shift + Right",
+            "click to store a location. Right click to recall",
+            "to that location. 10min CD."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "to that location. (CD: 10m)"
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 29]
```diff
@@ -0,0 +1,2620 @@
...
+        .addConsumer(TraitEventType.TICK, TickEvents::linksGhostly);
+
+    public static final MaterialTrait DYN_TESSERACT_HEAD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_HEAD)
+        .setAddedBy(SupportedPluginsManager.DYNATECH_NOTE)
+        .setTraitName("HyperCube [A]")
+        .setLore(
+            "Only works with both [A] and [B]. Shift + Right",
+            "click to store a location. Right click to recall",
+            "to that location. 10min CD."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "to that location. (CD: 10m)"
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 30]
```diff
@@ -0,0 +1,2620 @@
...
+        )
+        .addConsumer(TraitEventType.RIGHT_CLICK, RightClickEvents::headTessMat);
+
+    public static final MaterialTrait DYN_TESSERACT_PLATES = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_PLATES)
+        .setAddedBy(SupportedPluginsManager.DYNATECH_NOTE)
+        .setTraitName("Hyberbolic Tesseration")
+        .setLore(
+            "When in your personal home) you store energy.",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "When in your dimensional home, you store energy.",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 31]
```diff
@@ -0,0 +1,2620 @@
...
+        )
+        .addConsumer(TraitEventType.TICK, TickEvents::hyperbolic)
+        .addConsumer(TraitEventType.PLAYER_DAMAGED, PlayerDamagedEvents::hyperbolic);
+
+    public static final MaterialTrait DYN_TESSERACT_LINKS = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_LINKS)
+        .setAddedBy(SupportedPluginsManager.DYNATECH_NOTE)
+        .setTraitName("Hyberbolic Tesseration")
+        .setLore(
+            "When in your personal home) you store energy.",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "When in your dimensional home, you store energy.",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 32]
```diff
@@ -0,0 +1,2620 @@
...
+        .addConsumer(TraitEventType.RIGHT_CLICK, RightClickEvents::headTessMat);
+
+    public static final MaterialTrait DYN_TESSERACT_PLATES = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_PLATES)
+        .setAddedBy(SupportedPluginsManager.DYNATECH_NOTE)
+        .setTraitName("Hyberbolic Tesseration")
+        .setLore(
+            "When in your personal home) you store energy.",
+            "Damage taken is removed from this energy pool.",
+            "Required 8x pieces to function and caps at 50.",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Required 8x pieces to function and caps at 50.",
```

**Review by ybw0014 — CHANGES_REQUESTED** [review]
finished reviewing traits

**Comment by ybw0014** [comment]
I don't understand why my new reviews are outdated.

**Comment by Sefiraat** [comment]
> I don't understand why my new reviews are outdated.

I probably did something dumb :D - I can see them all so this is great, thanks!

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 33]
```diff
@@ -0,0 +1,2620 @@
...
+        .setLore(
+            "Does... nothing!"
+        );
+
+    public static final MaterialTrait CORE_IRON_ROD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_ROD)
+        .setAddedBy(SupportedPluginsManager.CORE_NOTE)
+        .setTraitName("Sharp 1")
+        .setLore(
+            "Gives increased damage) stacks with",
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "Gives increased damage, stacks with",
```

File: src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java [hunk 34]
```diff
@@ -0,0 +1,2620 @@
...
+        )
+        .addConsumer(TraitEventType.ENTITY_DAMAGED, EntityDamageEvents::linksIron);
+
+    public static final MaterialTrait CORE_GOLD_HEAD = new MaterialTrait()
+        .setPartType(MaterialTrait.PROP_HEAD)
+        .setAddedBy(SupportedPluginsManager.CORE_NOTE)
+        .setTraitName("Golden Veil")
+        .setLore(
+            "Makes the player invisible. If used on",
+            "a weapon) -100% damage."
```
Comment (ybw0014) [inline src/main/java/io/github/sefiraat/slimetinker/items/tinkermaterials/setup/Traits.java]:
```suggestion
            "a weapon, -100% damage."
```

**Review by ybw0014 — CHANGES_REQUESTED** [review]

**Comment by ybw0014** [comment]
It's ... outdated again.

**Review by ybw0014 — APPROVED** [review]
LGTM

**Comment by Sefiraat** [comment]
Nice, ill await the in-live feedback and merge tomorrow

**Pull request merged**